### PR TITLE
Convert unitest to pytest

### LINKED ISSUE #7: Massive test failures when used from Python 3.8
I am getting these failures when used from Python 3.8.3:

```
$ nosetests test
EE......E.FE.E.EE.E.EEEEEEEEEEEEE.EEEEEEEEEEEEEEEEEEEE.EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.EEEEEEEEEEEEEEEEEE.....SSSSSSSSSSSSSSSS.....................................S..............  C-c C-cS
======================================================================
ERROR: test.support.testresult.get_test_runner_class
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rocky/.pyenv/versions/3.8.3/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: get_test_runner_class() missing 1 required positional argument: 'verbosity'

======================================================================
ERROR: test.support.testresult.get_test_runner
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rocky/.pyenv/versions/3.8.3/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: get_test_runner() missing 2 required positional arguments: 'stream' and 'verbosity'

======================================================================
ERROR: test.audit-tests.test_open
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rocky/.pyenv/versions/3.8.3/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/rocky/.pyenv/versions/3.8.3/lib/python3.8/test/audit-tests.py", line 201, in test_open
    (open, sys.argv[2], "r"),
IndexError: list index out of range
-------------------- >> begin captured stdout << ---------------------
sys.addaudithook 139704741433248
... [hundreds of other lines]
```

I note however that when i use this package from 3.8.3 things work fine.

If you'd like me to investigate and suggest a fix/PR, let me know.

### PR DESCRIPTION
**Pull request by rocky**
Took a little longer than I expected. (However that just makes me appreciate the work that was done originally even more!)

One test that succeeded under unittest fails under pytest: 

```
def test_get():
    pq = pqdict(sample_items)
    assert pq.get("A") == 5
    assert pq.get("A", None) == 5
    assert pq.get("does_not_exist", None) is None

    # INVESTIGATE: in unit test, this succeded
    with pytest.raises(KeyError):
         pq.get('does_not_exist')
```

Fixes #7

### CODE REVIEW

**Comment by nvictus** [comment]
Interesting... pq.get is inherited from [Mapping](https://github.com/python/cpython/blob/3.8/Lib/_collections_abc.py#L641-L662) and maybe the behavior changed at some point. I would expect it to return `None`, just like it does now and just like `dict` does. I'm not sure why I expected it to raise KeyError before.

File: tests/test_pqdict.py [hunk 1]
```diff
@@ -4,467 +4,512 @@
...
+def test_values():
+    pq = pqdict(sample_items)
+    assert sorted(sample_values) == sorted(pq.values())
+    assert sorted(sample_values) == list(pq.popvalues())
+
+
+def test_items():
+    pq = pqdict(sample_items)
+    assert sorted(sample_items) == sorted(pq.items())
+    assert sorted(sample_values), [item[1] for item in pq.popitems()]
```
Comment (nvictus) [inline tests/test_pqdict.py:245]: I think you missed replacing `,` with `==` here.

File: tests/test_pqdict.py [hunk 2]
```diff
@@ -4,467 +4,512 @@
...
+    assert pq.precedes == operator.lt
+    pq = pqdict(reverse=True)
+    assert pq.precedes == operator.gt
+    func = lambda x, y: len(x) < len(y)
+    pq = pqdict(precedes=func)
+    pq["a"] = ()
+    pq["b"] = (1,)
+    pq["c"] = (1, 2)
+    pq["d"] = (1, 2, 3)
+    assert list(pq.popvalues()), [(), (1,), (1, 2) == (1, 2, 3)]
```
Comment (nvictus) [inline tests/test_pqdict.py:267]:
```suggestion
    assert list(pq.popvalues()) == [(), (1,), (1, 2), (1, 2, 3)]
```

File: tests/test_pqdict.py [hunk 3]
```diff
@@ -4,467 +4,512 @@
...
+        pq.top()
+    # non-empty
+    for num_items in range(1, 30):
+        items = generate_data("float", num_items)
         pq = pqdict(items)
-        for _ in range(100):
-            pq[random.choice(keys)] = random.randrange(25)
-            self._check_heap_invariant(pq)
-            self._check_index(pq)
+        assert pq.top(), min(items, key=lambda x: x[1])[0]
```
Comment (nvictus) [inline tests/test_pqdict.py:298]:
```suggestion
        assert pq.top() == min(items, key=lambda x: x[1])[0]
```

File: tests/test_pqdict.py [hunk 4]
```diff
@@ -4,467 +4,512 @@
...
+def test_topitem():
+    # empty
+    pq = pqdict()
+    with pytest.raises(KeyError):
+        pq.top()
+    # non-empty
+    for num_items in range(1, 30):
+        items = generate_data("float", num_items)
+        pq = pqdict(items)
+        assert pq.topitem(), min(items, key=lambda x: x[1])
```
Comment (nvictus) [inline tests/test_pqdict.py:318]:
```suggestion
        assert pq.topitem() == min(items, key=lambda x: x[1])
```

File: tests/test_pqdict.py [hunk 5]
```diff
@@ -4,467 +4,512 @@
...
+    assert keys_sorted[-1] == "bot"
+
+
+def test_datetime():
+    pq = pqdict()
+    dt = datetime.now()
+    pq["a"] = dt
+    pq["b"] = dt + timedelta(days=5)
+    pq["c"] = dt + timedelta(seconds=5)
+    assert list(pq.popkeys()), ["a", "c" == "b"]
```
Comment (nvictus) [inline tests/test_pqdict.py:497]:
```suggestion
    assert list(pq.popkeys()) == ["a", "c", "b"]
```

File: tests/test_pqdict.py [hunk 6]
```diff
@@ -4,467 +4,512 @@
...
+    mutable_value[0] = 0
+    assert pq[pq.top()] == [1]
+    pq.heapify("C")
+    assert pq[pq.top()] == [0]
+
+
+# The next test was originally in class TestModuleFunctions
+def test_nbest():
+    top3 = nlargest(3, dict(sample_items))
+    assert list(top3), ["F", "E" == "B"]
```
Comment (nvictus) [inline tests/test_pqdict.py:513]:
```suggestion
    assert list(top3) == ["F", "E", "B"]
```

File: tests/test_pqdict.py [hunk 7]
```diff
@@ -4,467 +4,512 @@
...
+    bot3 = nsmallest(3, dict(sample_items))
+    assert list(bot3), ["G", "D" == "A"]
```
Comment (nvictus) [inline tests/test_pqdict.py:515]:
```suggestion
    assert list(bot3) == ["G", "D", "A"]
```

File: tests/test_pqdict.py [hunk 8]
```diff
@@ -4,467 +4,512 @@
...
+    # TODO: FIX?
+    # pqdicts evaluate as equal even if they have different
+    # key functions and/or precedence functions
+    pq3 = maxpq(sample_items)
+    assert pq1 == pq3
+
+
+def test_minpq():
+    pq = minpq(A=5, B=8, C=7, D=3, E=9, F=12, G=1)
+    assert list(pq.popvalues()), [1, 3, 5, 7, 8, 9, 12]
```
Comment (nvictus) [inline tests/test_pqdict.py:96]:
```suggestion
    assert list(pq.popvalues()) == [1, 3, 5, 7, 8, 9, 12]
```

File: tests/test_pqdict.py [hunk 9]
```diff
@@ -4,467 +4,512 @@
...
+    assert pq.precedes == operator.lt
+
+
+def test_maxpq():
+    pq = maxpq(A=5, B=8, C=7, D=3, E=9, F=12, G=1)
+    assert list(pq.popvalues()), [12, 9, 8, 7, 5, 3, 1]
```
Comment (nvictus) [inline tests/test_pqdict.py:102]:
```suggestion
    assert list(pq.popvalues()) == [12, 9, 8, 7, 5, 3, 1]
```

File: tests/test_pqdict.py [hunk 10]
```diff
@@ -4,467 +4,512 @@
...
+    for val in iter(pq):
+        assert val in sample_keys
+    assert len(list(iter(pq))) == len(sample_keys)
+    assert len(pq) == n
+
+
+def test_keys():
+    pq = pqdict(sample_items)
+    assert sorted(sample_keys) == sorted(pq.keys())
+    assert sorted(sample_values), [pq[key] for key in pq.copy().popkeys()]
```
Comment (nvictus) [inline tests/test_pqdict.py:233]:
```suggestion
    assert sorted(sample_values) == [pq[key] for key in pq.copy().popkeys()]
```

File: tests/test_pqdict.py (hunk 1 above)
Comment (nvictus) [inline tests/test_pqdict.py:245]:
```suggestion
    assert sorted(sample_values) == [item[1] for item in pq.popitems()]
```

File: tests/test_pqdict.py [hunk 11]
```diff
@@ -4,467 +4,512 @@
...
+    pq.updateitem(key, value + 1.0)
+    assert pq[key] == value + 1.0
+    # can only update existing keys
+    with pytest.raises(KeyError):
+        pq.updateitem("does_not_exist", 99.0)
+
+
+def test_pushpopitem():
+    pq = minpq(A=5, B=8, C=1)
+    assert pq.pushpopitem("D", 10), "C" == 1
```
Comment (nvictus) [inline tests/test_pqdict.py:345]:
```suggestion
    assert pq.pushpopitem("D", 10) == ("C", 1)
```

File: tests/test_pqdict.py [hunk 12]
```diff
@@ -4,467 +4,512 @@
...
+    assert pq.pushpopitem("E", 5), "E" == 5
```
Comment (nvictus) [inline tests/test_pqdict.py:346]:
```suggestion
    assert pq.pushpopitem("E", 5) == ("E", 5)
```

File: tests/test_pqdict.py [hunk 13]
```diff
@@ -4,467 +4,512 @@
...
+# inherited implementations
+def test_get():
+    pq = pqdict(sample_items)
+    assert pq.get("A") == 5
+    assert pq.get("A", None) == 5
+    assert pq.get("does_not_exist", None) is None
+
+    # INVESTIGATE: in unit test, this succeded
+    # with pytest.raises(KeyError):
+    #     pq.get('does_not_exist')
```
Comment (nvictus) [inline tests/test_pqdict.py:190]: I believe `assert pq.get("does_not_exist") is None` should pass. I think this behavior is sensible and is consistent with `dict`. If you agree, let's change to that.

Though, I have no idea why the KeyError test used to pass before...

Comment (rocky) [reply tests/test_pqdict.py]: Yes, I agree. Should be changed in 38a5481

**Comment by rocky** [comment]
Thanks for reviewing. 

The think things are resolved now. I made those changes with a regexp search and replace which explains why some are faulty. 

(I make a lot of mistake which is why I work on debuggers. But testing is very important too!)

**Comment by nvictus** [comment]
Yeah, I figured it was a regexp issue.

Thank you so much for doing this! 

> I work on debuggers.

Very cool! I should check out some of your projects.

**Pull request merged**